📖 Provide a user manual (#2)
atlas-engine/Client.ts: README.md

- A client for accessing the `AtlasEngine`'s User domain
- A worker for processing External Tasks

## Quick example

```ts
import {UserDomainHttpClient} from '@atlas-engine/atlas_engine_client';

const atlasEngineUrl = 'http://localhost:44123';
const client = new UserDomainHttpClient(atlasEngineUrl);

const processDefinitions = await client.processDefinitions.getAll();
```

## Who should use it

This client is designed to let regular users access the `AtlasEngine`.

The ExternalTaskWorker lets you process External Tasks created by Service Tasks.

> TODO: Link to API docs once they are finished

## How to use the client

### Basic

You only need to provide an url to the `AtlasEngine` you want to access.
After that, the client is ready to use.

```ts
import {UserDomainHttpClient} from '@atlas-engine/atlas_engine_client';

// The URL of the AtlasEngine to connect to
const atlasEngineUrl = 'http://localhost:44123';

// Creating the client
const client = new UserDomainHttpClient(atlasEngineUrl);

// Querying all available UserTasks
const processDefinitions = await client.userTasks.query({}, myIdentity);
```

### With a client identity

All requests require an identity.

> By default, the client uses an internal identity that is auto-generated upon instantiaion.

Reviewer: ```suggestion
By default, the client uses an internal identity that is auto-generated upon instantiation.
```

However, this identity is only a dummy and will probably not have sufficient permissions to do what you need it to do.

Therefore, the client takes an identity as an optional second parameter:

```ts
import {IIdentity} from '@atlas-engine/iam.contracts';
import {UserDomainHttpClient} from '@atlas-engine/atlas_engine_client';

// The URL of the AtlasEngine to connect to
const atlasEngineUrl = 'http://localhost:8000';
// A custom identity
const myIdentity: IIdentity = {
  userId: 'myUserId',
  token: 'my-fancy-jwt-auth-token',
}

// Creating the client
const client = new UserDomainHttpClient(atlasEngineUrl, myIdentity);

// Querying all available UserTasks

> const userTasks = await client.userTasks.query({});

Author: ```suggestion
const userTasks = await client.userTasks.query();
```

> const userTasks = await client.userTasks.query({}, myIdentity);

Reviewer: Ich würde hier die myIdentity rausnehmen. Dann ist es klarer dass man die Identity für den ganzen Client vorher als Default gesetzt hat und es nun nicht für jeden Request braucht. 

```suggestion
const processDefinitions = await client.userTasks.query({});
```

// The URL of the AtlasEngine to connect to
const atlasEngineUrl = 'http://localhost:8000';
// The topic for which to process External Tasks
const externalTaskTopic = 'my_fancy_topic';

// A handler function for processing the External Tasks
const externalTaskHandler = <TPayload, TResult>(externalTask: ExternalTask<TPayload>): TResult => {
  // Do something here
  return {hello: 'world'};
}

// Creating the worker
const externalTaskWorker = new ExternalTaskWorker<TPayload, TResult>(
  atlasEngineUrl,
  externalTaskTopic,
  externalTaskHandler,
);
```

Note that each worker can only subscribe to a single topic.

### With extended configuration

You can provide a number of optional configuration parameters to a worker.

These include:
- `lockDuration`: The amount of time by which an External Task should be locked for the worker. Defaults to 30 seconds
- `maxTasks`: The max number of tasks that the worker should process simultaneously. Defaults to 10
- `longpollingTimeout`: The amount of time that the worker should wait for External Tasks to become available with each request. Defaults to 10 seconds
- `identity`: The identity to use for the worker

> As with the User domain client, it is highly recommended - but not mandatory - to provide an identity here, since the dummy identity

Reviewer: ```suggestion
As with the UserDomainClient, it is highly recommended - but not mandatory - to provide an identity here, since the dummy identity
```

oder

```suggestion
As with the user domain client, it is highly recommended - but not mandatory - to provide an identity here, since the dummy identity
```
